crates/tracing-instrumentation/src/prometheus_metrics.rs

```diff
@@ -74,7 +81,23 @@ impl Prometheus {
     }
 
     pub fn global_labels(&self) -> &Vec<String> {
-        &self.global_labels
+        if let Some(global_labels) = self.global_labels.get() {
+            return global_labels;
+        }
+
+        let Some(node_id) = GLOBAL_NODE_ID.get() else {
```

Comment:
why didn't initialize it at `install` method？

Reply:
Unfortunately, node_id is not known at that time - we only learn it after we've contacted the metadata service. See:

https://github.com/restatedev/restate/blob/41e606e6c2c6535134f0e7ee5d6bd70ce12b924a/crates/node/src/init.rs#L111-L113

We could defer Prometheus init until after that's done, too; but we start Prometheus as early as possible and already had this in place for tracing so I reused it.

Reply:
> Unfortunately, node_id is not known at that time - we only learn it after we've contacted the metadata service. See:
> 
> https://github.com/restatedev/restate/blob/41e606e6c2c6535134f0e7ee5d6bd70ce12b924a/crates/node/src/init.rs#L111-L113
> 
> We could defer Prometheus init until after that's done, too; but we start Prometheus as early as possible and already had this in place for tracing so I reused it.

ok，thanks for you further explanation. Can we just update the whole global_labels when we got the node_id from metadata-server？I think it's a bit strange to perform the judgment and assignment when retrieving.

Reply:
So something more like what I had in the first commit? https://github.com/restatedev/restate/pull/3027/commits/cc8234bfd03283e21b18d02125f89c2c383ccf1a

I was hoping to remove the clone downstream altogether and not re-create the labels once computed. But I see that due to the way make the cf-specific metrics in `render_metrics` / `metrics_exporter_prometheus::formatting::write_metric_line`, we'll probably always have to clone. So yeah, it might just be worth reverting to the earlier version.